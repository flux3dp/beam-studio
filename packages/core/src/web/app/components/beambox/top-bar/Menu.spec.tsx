import React from 'react';

import { fireEvent, render } from '@testing-library/react';

const defaultGlobalPreference = {
  'anti-aliasing': true,
  auto_align: true,
  'enable-uv-print-file': true,
  show_grids: true,
  show_rulers: true,
  use_layer_color: true,
  zoom_with_window: true,
};
const mockGlobalPreference = { ...defaultGlobalPreference };
const mockUseGlobalPreferenceStore = jest.fn();

jest.mock('@core/app/stores/globalPreferenceStore', () => ({
  useGlobalPreferenceStore: mockUseGlobalPreferenceStore,
}));

const open = jest.fn();

jest.mock('@core/implementations/browser', () => ({
  open,
}));

const mockDiscoverRemoveListener = jest.fn();

jest.mock('@core/helpers/api/discover', () => () => ({
  removeListener: mockDiscoverRemoveListener,
}));

const emit = jest.fn();
const on = jest.fn();
const removeListener = jest.fn();

jest.mock('@core/helpers/eventEmitterFactory', () => ({
  createEventEmitter: () => ({
    emit,
    on,
    removeListener,
  }),
}));

window.os = 'MacOS';

import Menu from './Menu';

describe('should render correctly', () => {
  beforeEach(() => {
    Object.assign(mockGlobalPreference, defaultGlobalPreference);
    mockUseGlobalPreferenceStore.mockImplementation((selector) => selector(mockGlobalPreference));
  });

  test('open the browser and reach the correct page', () => {
    mockUseGlobalPreferenceStore.mockReturnValue(true);

    const { container, getByText } = render(<Menu email={undefined} />);

    expect(container).toMatchSnapshot();

    fireEvent.click(container.querySelector('div.menu-btn-container'));
    expect(container).toMatchSnapshot();

    fireEvent.click(getByText('Help'));
    expect(container).toMatchSnapshot();

    fireEvent.click(getByText('Help Center'));
    expect(open).toHaveBeenCalledTimes(1);
    expect(open).toHaveBeenNthCalledWith(1, 'https://helpcenter.flux3dp.com/');
  });

  test('test checkbox menu item', () => {
    const keys = Object.keys(mockGlobalPreference);

    for (const key of keys) {
      mockGlobalPreference[key] = false;
    }

    const { container, getByText, rerender } = render(<Menu />);

    expect(container).toMatchSnapshot();

    fireEvent.click(container.querySelector('div.menu-btn-container'));
    fireEvent.click(getByText('View'));
    expect(container).toMatchSnapshot();

    fireEvent.click(getByText('Show Rulers'));
    expect(emit).toHaveBeenCalledTimes(1);
    expect(emit).toHaveBeenNthCalledWith(1, 'MENU_CLICK', null, {
      id: 'SHOW_RULERS',
    });
    mockGlobalPreference.show_rulers = true;
    rerender(<Menu />);
    expect(container).toMatchSnapshot();
  });

  test('already signed in', () => {
    mockUseGlobalPreferenceStore.mockReturnValue(true);

    const { container, getByText } = render(<Menu email="tester@flux3dp.com" />);

    fireEvent.click(container.querySelector('div.menu-btn-container'));
    fireEvent.click(getByText('Account'));
    expect(container).toMatchSnapshot();
  });
});
