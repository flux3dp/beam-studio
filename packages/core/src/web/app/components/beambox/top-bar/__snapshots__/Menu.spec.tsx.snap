// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly already signed in 1`] = `
<div>
  <div
    class="menu-btn-container"
  >
    <img
      class="icon"
      src="img/logo-line.svg"
    />
    <img
      class="icon-arrow"
      src="img/icon-arrow-d.svg"
    />
  </div>
  <div
    class="rc-menu-container"
  >
    <ul
      aria-label="Menu"
      class="rc-menu rc-menu--open rc-menu--animation rc-menu--dir-bottom"
      role="menu"
      style="left: 0px; top: 0px;"
      tabindex="-1"
    >
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          File
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Edit
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          View
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Machines
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Tools
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="true"
          aria-haspopup="true"
          class="rc-menu__item rc-menu__item--open"
          role="menuitem"
          tabindex="-1"
        >
          Account
        </div>
        <ul
          aria-label="Account"
          class="rc-menu rc-menu--open rc-menu--animation rc-menu--dir-right"
          role="menu"
          style="left: 0px; top: 0px;"
          tabindex="-1"
        >
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Sign Out (tester@flux3dp.com)
          </li>
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Design Market
          </li>
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Manage My Account
          </li>
        </ul>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Help
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`should render correctly open the browser and reach the correct page 1`] = `
<div>
  <div
    class="menu-btn-container"
  >
    <img
      class="icon"
      src="img/logo-line.svg"
    />
    <img
      class="icon-arrow"
      src="img/icon-arrow-d.svg"
    />
  </div>
  <div
    class="rc-menu-container"
  />
</div>
`;

exports[`should render correctly open the browser and reach the correct page 2`] = `
<div>
  <div
    class="menu-btn-container"
  >
    <img
      class="icon"
      src="img/logo-line.svg"
    />
    <img
      class="icon-arrow"
      src="img/icon-arrow-d.svg"
    />
  </div>
  <div
    class="rc-menu-container"
  >
    <ul
      aria-label="Menu"
      class="rc-menu rc-menu--open rc-menu--animation rc-menu--dir-bottom"
      role="menu"
      style="left: 0px; top: 0px;"
      tabindex="-1"
    >
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          File
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Edit
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          View
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Machines
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Tools
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Account
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Help
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`should render correctly open the browser and reach the correct page 3`] = `
<div>
  <div
    class="menu-btn-container"
  >
    <img
      class="icon"
      src="img/logo-line.svg"
    />
    <img
      class="icon-arrow"
      src="img/icon-arrow-d.svg"
    />
  </div>
  <div
    class="rc-menu-container"
  >
    <ul
      aria-label="Menu"
      class="rc-menu rc-menu--open rc-menu--animation rc-menu--dir-bottom"
      role="menu"
      style="left: 0px; top: 0px;"
      tabindex="-1"
    >
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          File
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Edit
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          View
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Machines
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Tools
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Account
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="true"
          aria-haspopup="true"
          class="rc-menu__item rc-menu__item--open"
          role="menuitem"
          tabindex="-1"
        >
          Help
        </div>
        <ul
          aria-label="Help"
          class="rc-menu rc-menu--open rc-menu--animation rc-menu--dir-right"
          role="menu"
          style="left: 0px; top: 0px;"
          tabindex="-1"
        >
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            About Beam Studio
          </li>
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Show Start Tutorial
          </li>
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Show Interface Introduction
          </li>
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Hand Gesture Introduction
          </li>
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Feedback Questionnaire
          </li>
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Change Logs
          </li>
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Help Center
          </li>
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Keyboard shortcuts
          </li>
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Contact Us
          </li>
          <li
            class="rc-menu__divider"
            role="separator"
          />
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Community Forum
          </li>
          <li
            class="rc-menu__item"
            role="menuitem"
            tabindex="-1"
          >
            Follow Us
          </li>
        </ul>
      </li>
    </ul>
  </div>
</div>
`;

exports[`should render correctly test checkbox menu item 1`] = `
<div>
  <div
    class="menu-btn-container"
  >
    <img
      class="icon"
      src="img/logo-line.svg"
    />
    <img
      class="icon-arrow"
      src="img/icon-arrow-d.svg"
    />
  </div>
  <div
    class="rc-menu-container"
  />
</div>
`;

exports[`should render correctly test checkbox menu item 2`] = `
<div>
  <div
    class="menu-btn-container"
  >
    <img
      class="icon"
      src="img/logo-line.svg"
    />
    <img
      class="icon-arrow"
      src="img/icon-arrow-d.svg"
    />
  </div>
  <div
    class="rc-menu-container"
  >
    <ul
      aria-label="Menu"
      class="rc-menu rc-menu--open rc-menu--animation rc-menu--dir-bottom"
      role="menu"
      style="left: 0px; top: 0px;"
      tabindex="-1"
    >
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          File
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Edit
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="true"
          aria-haspopup="true"
          class="rc-menu__item rc-menu__item--open"
          role="menuitem"
          tabindex="-1"
        >
          View
        </div>
        <ul
          aria-label="View"
          class="rc-menu rc-menu--open rc-menu--animation rc-menu--dir-right"
          role="menu"
          style="left: 0px; top: 0px;"
          tabindex="-1"
        >
          <li
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitem"
            tabindex="-1"
          >
            <span
              class="action"
            >
              Zoom In
            </span>
            <span
              class="hotkey"
            >
              ⌘+
            </span>
          </li>
          <li
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitem"
            tabindex="-1"
          >
            <span
              class="action"
            >
              Zoom Out
            </span>
            <span
              class="hotkey"
            >
              ⌘-
            </span>
          </li>
          <li
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitem"
            tabindex="-1"
          >
            Fit to Window Size
          </li>
          <li
            aria-checked="false"
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Auto Fit to Window Size
          </li>
          <li
            class="rc-menu__divider"
            role="separator"
          />
          <li
            aria-checked="false"
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Show Grids
          </li>
          <li
            aria-checked="false"
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Show Rulers
          </li>
          <li
            aria-checked="false"
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Use Layer Color
          </li>
          <li
            aria-checked="false"
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Snap To Vertices
          </li>
          <li
            aria-checked="false"
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Anti-Aliasing
          </li>
        </ul>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Machines
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Tools
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Account
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Help
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`should render correctly test checkbox menu item 3`] = `
<div>
  <div
    class="menu-btn-container"
  >
    <img
      class="icon"
      src="img/logo-line.svg"
    />
    <img
      class="icon-arrow"
      src="img/icon-arrow-d.svg"
    />
  </div>
  <div
    class="rc-menu-container"
  >
    <ul
      aria-label="Menu"
      class="rc-menu rc-menu--closing rc-menu--animation rc-menu--dir-bottom"
      role="menu"
      style="left: 0px; top: 0px;"
      tabindex="-1"
    >
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          File
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Edit
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          View
        </div>
        <ul
          aria-label="View"
          class="rc-menu rc-menu--closing rc-menu--animation rc-menu--dir-right"
          role="menu"
          style="left: 0px; top: 0px;"
          tabindex="-1"
        >
          <li
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitem"
            tabindex="-1"
          >
            <span
              class="action"
            >
              Zoom In
            </span>
            <span
              class="hotkey"
            >
              ⌘+
            </span>
          </li>
          <li
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitem"
            tabindex="-1"
          >
            <span
              class="action"
            >
              Zoom Out
            </span>
            <span
              class="hotkey"
            >
              ⌘-
            </span>
          </li>
          <li
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitem"
            tabindex="-1"
          >
            Fit to Window Size
          </li>
          <li
            aria-checked="false"
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Auto Fit to Window Size
          </li>
          <li
            class="rc-menu__divider"
            role="separator"
          />
          <li
            aria-checked="false"
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Show Grids
          </li>
          <li
            aria-checked="true"
            class="rc-menu__item rc-menu__item--type-checkbox rc-menu__item--checked"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Show Rulers
          </li>
          <li
            aria-checked="false"
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Use Layer Color
          </li>
          <li
            aria-checked="false"
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Snap To Vertices
          </li>
          <li
            aria-checked="false"
            class="rc-menu__item rc-menu__item--type-checkbox"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            Anti-Aliasing
          </li>
        </ul>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Machines
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Tools
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Account
        </div>
      </li>
      <li
        class="rc-menu__submenu"
        role="presentation"
      >
        <div
          aria-expanded="false"
          aria-haspopup="true"
          class="rc-menu__item"
          role="menuitem"
          tabindex="-1"
        >
          Help
        </div>
      </li>
    </ul>
  </div>
</div>
`;
